### ISSUE TITLE
Wrong patch is generated when siblings exists on several levels.

### ISSUE THREAD

**Issue by chrhol** [Starcounter/Starcounter#3680]
Following the instruction in https://github.com/StarcounterPrefabs/Launcher/issues/229#issue-153717013 somehow a patch for the second workspace comes with a path that points to the wrong place. 

Exactly why this happens I'm not sure about, but it has to do with siblings (or maybe even the same sibling) appearing in different places in the viewmodel. 

**Note**: to reproduce it following the steps in the other issue, the SignIn-app needs to be compiled with the version from the https://github.com/StarcounterSamples/SignIn/commit/7cd21a11b087567d45a17b09f2dd72707c6a31fd commit

**Comment by chrhol** [comment]
I'm not sure anymore that it's the patches that is the problem. Something strange is going on with the siblings and merging. I added debugcode in Json that is numbering the json-instances, to be able to distinguish them, and exposed it so it's sent to the client. 

In the following snippet (I removed all other properties), the exact same siblings for warehouse, launcher and signin exists in *both* workspaces, though on different levels:
```javascript
{
  "workspaces": [
    { // workspace 0
      "Launcher": {
        "InstanceNo": 150,
      },
      "Warehouse": {
        "InstanceNo": 152,
        "Partial": {
          "Warehouse": {
            "InstanceNo": 182, // exists in other ws
          },
          "Launcher": {
            "InstanceNo": 173, // exists in other ws
          },
          "SignIn": {
            "InstanceNo": 161, // exists in other ws
          }
        },
      }
    }, // end workspace 0
    { // workspace 1
      "Launcher": {
        "InstanceNo": 173, // exists in other ws
      },
      "Warehouse": {
        "InstanceNo": 182, // exists in other ws
      },
      "SignIn": {
        "InstanceNo": 161, // exists in other ws
      }
    } // workspace 1
  ],
}
```

This makes it impossible in the current implementation to know which path is the intended one. We had discussions before if we should support having the same sibling in different places in the viewmodel and generate patches for all paths, but decided at that time that it wasn't needed. If we are going to support this, it will mean a quite big rewrite of current jsonpatch implementation, and not feasible to fix before the next stable.

So, we have to make a decision here. Either we should support having the same partial in different places and when generating the patches this needs to be taken in consideration OR we don't allow it and need to add some check in the jsonmerger that throws an exception in these cases.

Not sure which way to take, @miyconst, @warpech, @dan31

**Comment by dan31** [comment]
I will also CC @tomalec for that; what do you think?

**Comment by miyconst** [comment]
I honestly believed that I have answered this issue, apparently not :(.

The source of this problem is impossibility to update session-less partials. The SignIn app attaches a single page to `Session.Current.Data` and returns it in every place it's needed. This is required to be able to access the page from a commit hook and update it once user sign in state has been changed.

It is also not possible to attach multiple partials per session to response with a unique partial per request. It is possible to save collection of partials inside a root partial, but there is no way to know when to remove a no longer used partial.

The only possible solution seem to be bound properties. Instead of updating partials from commit hooks, the partials will be self updated. In this case the update code will be executed every time view-model changes collected. Some caching can be added to avoid expensive calculations, but it's still not the perfect solution.

**Comment by miyconst** [comment]
Another very major issue found. Once a partial being involved in different places of the JSON tree, client side changes to some of the parts are being ignored by server. Means the `Handle` methods are not being called.

We really have to do something about it.

**Comment by tomalec** [comment]
I'm not very familiar with our use cases for multiple insertion points of same partial.

However, I think it would be really hard to avoid them with extensive use of mappings. I'm afraid it would be hard to demand that person writing (generic) mapper for app A will be aware of all the other mappers running by an instance for other apps.

Forgive me, if I didn't get where the root cause of a problem is, but I believe that in long-term solution, the only constraint we should force is to avoid circular references.

On the other hand, having same object referenced in many places of the same JSON tree is not possible, so implementing such feature would be our piece of magic, but as we can achieve "same object" by going through entire Puppet flow, it makes sense to me. 
> It's not the same object, it's another representation of the same object"

As we can refer the same Person in many places why can we refer same view-model in many places?

**Comment by miyconst** [comment]
>Forgive me, if I didn't get where the root cause of a problem is, but I believe that in long-term solution, the only constraint we should force is to avoid circular references.

You are absolutely right, with mapping it won't be possible to avoid. I described the reason why it happened in SignIn with a single mapping.

**Comment by chrhol** [comment]
On the longterm solution I agree, we should support it, but my point was that it will require a rewrite on how patches are created on the server. And it's not feasible to have this change in the next stable version.

The thing that makes it a bit different, is that we have the same instance on the server, but it's treated as separate instances in the viewmodel on the client. Which means that it's not enough to just create one patch for one change, we need to evaluate all possible paths and and create patches for them all. 

One serverside change -> one or many clientside changes. 

> Another very major issue found. Once a partial being involved in different places of the JSON tree, client side changes to some of the parts are being ignored by server. Means the Handle methods are not being called.

> We really have to do something about it.

This doesn't sound like its the same problem. This is most likely a separate bug.

I'm not sure I agree on the showstopper label either. Granted we should fix it, but it's hardly something critical. It's been this way since the beginning and we got no other issues reported than this with the sign-in, that isn't even in the latest version of the sign-in app (you have to check out an older version to reproduce it)

But the conclusion is then, that we want to support this behaviour and we need to handle so that all possible paths are patched. This is a bigger change (not simply a bugfix) and I would really recommend to *not* have this change in the next stable version.

**Comment by chrhol** [comment]
>> Another very major issue found. Once a partial being involved in different places of the JSON tree, client side changes to some of the parts are being ignored by server. Means the Handle methods are not being called.

>> We really have to do something about it.

> This doesn't sound like its the same problem. This is most likely a separate bug.

Actually, what it could be is that the siblings gets messed up when the same instance gets merged more then once so the correct sibling/parent is not found when evaluating the patch.

**Comment by chrhol** [comment]
I will start looking into this a bit later today.

**Comment by chrhol** [comment]
Thanks for the feedback btw :+1:

**Comment by miyconst** [comment]
@chrhol I have replaced the `showtopper` label with `High prio`, even if we do not provide the proper support in the next stable, we should redesign our apps to don't save single partial in session and response in different places.

Right now we use mapping in MM very intensively and sooner or later it will start to hit back.

>Actually, what it could be is that the siblings gets messed up when the same instance gets merged more then once so the correct sibling/parent is not found when evaluating the patch.

I was wrong saying that the changes coming from the client side are ignored. The `Handle` method is also called correctly. But the changes made in the method are not send back to the client. 

This issue already hits me. Latest SignIn, latest Starcounter. Steps to reproduce:

- Run MM solution.
- Open `/launcher`.
- SignIn with admin/admin.
- SignOut.
- Try to SignIn again.
- The button gets disabled once patch sent to the server.
- The `void Handle(Input.SignInClick Action)` method in `SignIn/ViewModels/SignInFormPage.json.cs` gets hit.
- Updating patch from the server never get back.

**Comment by chrhol** [comment]
> This issue already hits me. Latest SignIn, latest Starcounter. Steps to reproduce:

This is another problem and has to do with having `SessionContainer` as only a code-behind file, bypassing creating a template and keeping track of changes properly. I will add a new issue to discuss how to handle that. If you change this class to contain a JSON-by-example file containing the two properties `SignIn` and `Master` instead of having it as only properties in the code-behind it will work.

**Comment by miyconst** [comment]
@chrhol thank you for describing this. I have added the JSON template for the `SessionContainer` and it works!

**Comment by chrhol** [comment]
I'm currently working on this issue. I have made some progress, but there's still some way to go to have it work properly.

**Comment by chrhol** [comment]
This is going sloooow. The actual creation of patches is not a major obstacle. My main problem, and what got me stuck is the jsonmerger. With the changes needed internally to support the pattern of having the same jsoninstance in several places, it's becoming more and more obvious that the current version of how and when the merger is called is not working.

The conclusion is that #3538 is needed (or some version of that) to finalize this issue.

I will switch over to test the changes in #3538 and also write a document that will (or at least try to) outline how the jsonmerger and the callback for the merger works.
